### PR TITLE
Add typing for `sync`

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #xxxx

- [ ] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/utils.py [hunk 1]
```diff
@@ -371,17 +371,27 @@ def in_async_call(loop, default=False):
...
         raise RuntimeError("IOLoop is closed")
 
     e = threading.Event()
     main_tid = threading.get_ident()
-    result = error = future = None  # set up non-locals
+
+    # set up non-locals
+    result: T
+    error = None
+    future: asyncio.Task
```
Comment (crusaderky) [inline distributed/utils.py:394]:
```suggestion
    future: asyncio.Future
```
if func doesn't return a coroutine, ensure_future won't return a Task

File: distributed/utils.py [hunk 2]
```diff
@@ -371,17 +371,27 @@ def in_async_call(loop, default=False):
         return False
 
 
-def sync(loop, func, *args, callback_timeout=None, **kwargs):
+def sync(
+    loop: IOLoop,
+    func: Callable[..., Awaitable[T]],
+    *args: AnyType,
+    callback_timeout: str | float | timedelta | None = None,
+    **kwargs: AnyType,
```
Comment (crusaderky) [inline distributed/utils.py:379]:
```suggestion
    func: Callable[P, Awaitable[T]],
    *args: P.args,
    callback_timeout: str | float | timedelta | None = None,
    **kwargs: P.kwargs,
```

add at the top of the module:
```python
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    # TODO import from typing (requires Python >=3.10)
    from typing_extensions import ParamSpec

    P = ParamSpec("P")
```

**Review by crusaderky — CHANGES_REQUESTED** [review]
.

File: distributed/utils.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/utils.py:379]: From what I understand, `ParamSpec` doesn't work in combination with additional keyword arguments (`callback_timeout`). At least, that's what I gathered from the PEP and trying to get this to work. LMK if you know how to pull this off!

File: distributed/utils.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/utils.py]: Fair point, the typing here is kinda awkward so I had ignored this for simplicity. I've adjusted this to be as accurate as possible without `overload`ing `sync`.

File: distributed/utils.py [hunk 3]
```diff
@@ -371,17 +371,27 @@ def in_async_call(loop, default=False):
         return False
 
 
-def sync(loop, func, *args, callback_timeout=None, **kwargs):
+def sync(
+    loop: IOLoop,
+    func: Callable[..., Awaitable[T] | asyncio.Future[T]],
```
Comment (graingert) [inline distributed/utils.py:376]: Future[T] is already an Awaitable[T] so this union is redundant

**Review by graingert — CHANGES_REQUESTED** [review]

File: distributed/utils.py (hunk 2 above)
Comment (graingert) [reply distributed/utils.py]: You can get this working with just the *args with TypeVarTuple and Unpack but you'll need to opt into that support on mypy 1.6

This TypeVarTuple and Unpack will show up as Any to users without this support so it's safe to opt in here

File: distributed/utils.py [hunk 4]
```diff
@@ -371,17 +371,27 @@ def in_async_call(loop, default=False):
...
     main_tid = threading.get_ident()
-    result = error = future = None  # set up non-locals
+
+    # set up non-locals
+    result: T
+    error = None
+    future: asyncio.Future[T] | asyncio.Task[T]
 
     @gen.coroutine
     def f():
```
Comment (graingert) [inline distributed/utils.py:397]:
```suggestion
    def f() -> Generator[Any, Any, None]:
```

File: distributed/utils.py [hunk 5]
```diff
@@ -412,14 +422,14 @@ def wait(timeout):
             raise
 
     loop.add_callback(f)
```
Comment (graingert) [inline distributed/utils.py:424]: def cancel and def wait are missing annotations

File: distributed/utils.py [hunk 6]
```diff
@@ -371,29 +371,39 @@ def in_async_call(loop, default=False):
...
         raise RuntimeError("IOLoop is closed")
 
     e = threading.Event()
     main_tid = threading.get_ident()
-    result = error = future = None  # set up non-locals
+
+    # set up non-locals
+    result: T
+    error = None
+    future: asyncio.Future[T] | asyncio.Task[T]
```
Comment (graingert) [inline distributed/utils.py:394]: Task is an instance of Future so 
```suggestion
    future: asyncio.Future[T]
```

File: distributed/utils.py [hunk 7]
```diff
@@ -371,29 +371,39 @@ def in_async_call(loop, default=False):
...
 
     @gen.coroutine
-    def f():
+    def f() -> Generator[Any, Any, None]:
```
Comment (graingert) [inline distributed/utils.py]: Argh I fell into the trap in this file
```suggestion
    def f() -> Generator[AnyType, AnyType, None]:
```

Comment (hendrikmakait) [reply distributed/utils.py]: Fixed already :)

File: distributed/utils.py (hunk 6 above)
Comment (hendrikmakait) [reply distributed/utils.py:394]: Is it? The docs state
> [asyncio.Task](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task) inherits from [Future](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future) all of its APIs except [Future.set_result()](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.set_result) and [Future.set_exception()](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.set_exception).

...but I guess it's close enough for our purposes.

File: distributed/utils.py (hunk 5 above)
Comment (hendrikmakait) [reply distributed/utils.py:424]: Addressed.

File: distributed/utils.py (hunk 3 above)
Comment (hendrikmakait) [reply distributed/utils.py]: Addressed.

File: distributed/utils.py [hunk 8]
```diff
@@ -431,6 +431,7 @@ def wait(timeout):
 
     if error is not None:
         typ, exc, tb = error
+        assert exc is not None
```
Comment (graingert) [inline distributed/utils.py:434]: Would this be redundant if you type annotated error on line 393 as

`error: tuple[type[BaseException], BaseException, types.TracebackType] | None = None`
?

Also `sys.exc_info()` is pending deprecation in favour of getting the info from `type(e), e, e.__traceback__`

Comment (hendrikmakait) [reply distributed/utils.py:434]: I've found a variant that works if I avoid `sys.exc_info()`.

Comment (hendrikmakait) [reply distributed/utils.py:434]: Yours is cleaner, will adjust.

Comment (graingert) [reply distributed/utils.py:434]: Will have to check how the traceback appears in the console log with these different options

File: distributed/utils.py (hunk 6 above)
Comment (hendrikmakait) [reply distributed/utils.py]: For papertrail: https://github.com/python/typeshed/blob/main/stdlib/asyncio/tasks.pyi#L382

Typing considers them the same.

File: distributed/utils.py (hunk 8 above)
Comment (hendrikmakait) [reply distributed/utils.py]: It looks like a simple `raise error` works as expected.

Comment (graingert) [reply distributed/utils.py]: Awesome

File: distributed/utils.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/utils.py:379]: Let's upgrade mypy and utilize `TypeVarTuple` in a follow-up PR.

**Pull request merged**